### DISCUSSION TITLE
Migration of third party fields to JSON via BaseContentRefactorMigration

### DISCUSSION THREAD

**Discussion by rynpsc** [craftcms/cms#14017] [General]
With the move to storing content as JSON in Craft 5, Craft has the `BaseContentRefactorMigration` and `m230511_215903_content_refactor.php` for handling the conversion of native fields, in particular the `decodeValue` function. Would it be possible for third-party fields to take advantage of this migration without needing to create their own migration path and duplicate the code?

The context for this is that I have a field type that is currently storing a JSON value as text using `Json::encode`. It would be nice if where Craft handles its native field types (like Table fields), third parties could update theirs as well to store the value as JSON.

https://github.com/craftcms/cms/blob/97b88b18c6b4f169e376660b40564b3014e5f78d/src/migrations/BaseContentRefactorMigration.php#L231-L236

Could this be done via an event within `BaseContentRefactorMigration`? Or maybe that's not required and Craft could run anything that doesn't match the `switch` statement through `decodeIfJson`?

https://github.com/craftcms/cms/blob/97b88b18c6b4f169e376660b40564b3014e5f78d/src/migrations/BaseContentRefactorMigration.php#L242-L257

In my case this is a nice to have and a migration path isn't strictly required and as I can continue to support JSON encoded strings for backwards compatibility and start storing new values as JSON (or just get people to re-save all elements - maybe that will be required as part of the Craft 5 upgrade path, and makes this discussion redundant?) but it would be nice to normalise things.

**Comment by brandonkelly** [comment]
Thanks for bringing that up!

I’ve just updated the migration to start JSON-decoding values for fields whose `dbType()` returns `yii\db\Schema::TYPE_JSON`. And the migration no longer makes a special case for Table or Lightswitch fields. (64f9b746bc70347380c77eaed8d4dac20605615a)

So to get the migration to automatically JSON-decode your field’s values, add the following to it:

```php
public static function dbType(): array|string|null
{
    return Schema::TYPE_JSON;
}
```

**Reply by rynpsc** [reply]
Gave that a test and it's handled the migration perfectly.

However, it does mean it's no longer possible to pass a key to `valueSql` as per https://github.com/craftcms/cms/discussions/14040#discussioncomment-7858724 which I had adopted to provide a query condition rule on a field that stores JSON.

I've also opened https://github.com/craftcms/cms/issues/14054 regarding some issues I was running into with migrations prior to https://github.com/craftcms/cms/commit/64f9b746bc70347380c77eaed8d4dac20605615a.

I could continue to return an array from `dbType` to support querying the JSON by key, and handle the migration myself but https://github.com/craftcms/cms/issues/14054 means that currently returning an array from `dbType` breaks the migration.

Would the best way to proceed be to return `Schema::TYPE_JSON` from `dbType` and then add the ability to pass a type along with the key in `valueSql` to override to type casting in `getValueSql` where it checks for the key being returned from `dbType`? (Doesn't address returning an array currently breaking the migration/incorreclty changing the stored value)

In an ideal world there would be a way to:

- convert JSON encoded strings to JSON within the Craft 5 migration, either by Craft itself or by the developer
- be able to query JSON values by key using Craft's `jsonExtract` method, requiring support from `valueSql` and `getValueSql`

**Related**

https://github.com/craftcms/cms/issues/14054
https://github.com/craftcms/cms/issues/14067
https://github.com/craftcms/cms/discussions/14040

**Reply by brandonkelly** [reply]
Sorry, overlooked that! I’ve just made some changes so that fields whose `dbType()` method returns an array will be JSON-decoded, if there weren’t multiple field columns in Craft 4. (066b050a9e2e1690f1b79637e82943e3ffa99548)

**Reply by brandonkelly** [reply]
Craft 5.0.0-alpha.3 is out now with those changes.

**Reply by rynpsc** [reply]
Thanks Brandon, everything is now working 🎉
